### DISCUSSION TITLE
hasProp, hasPropSatisfying

### DISCUSSION THREAD

**Discussion by airjp73** [remeda/remeda#959] [Suggest a Function]
I often find myself wanting to `filter` an array based on whether or not a property exists or that has a property of a certain value (like when working with discriminated unions).

I'm wondering if there's interest in adding some helpers like `hasNonNilProperty` or `hasPropertyWithValue` to the library. I'm happy to work on a PR if this seems like something reasonable to add.

~~I made a [Typescript Playground](https://www.typescriptlang.org/play?#code/PQkAgUQDwQwWwA4BsCmBnMB7AZmBAnFAEwEsBjGAF3TEoAsqwB3TAVySLACMUwA7cr0qYwDAG686KJAhT4M2TPgBQyygE9ZYACqYiIgLxhlYMAG8waSlVZoAXGADkZTIlTUijgDRgXblB4AgpQOACJUKD7UUCGWlPgkfADmPqwIRBFEAELqAPwOVgnJYAC+JmAAPuZxNvZOfJiUALR+yAHE3rQoMQXxiSlgaRkeOflxRUmlqi58VrR6mBhGANoAumAwGLr6awDcyjNzre7E24tgRsL6aOUAdNgkSNT4ABQvV5gAlA4fYCQYGlkOHm+jAADJzFZag5nK42h5HCULgA+EGYW5Qyi2C4GIyw-wIz4HTCzSi+OEnIjBC7kgmnBZoZYABlWt2O7SplGU4FMYAAerlpiS5h80AB1Ej0ACq6UyAAV8JhZPgNDTRXcHk85G8Pt80X8AZoUMDfhCAEooACOrBIhCIAB45eQANb2wHG3AfHyOIaZHKOZGogyogCEIY+t19I3URMOZKjxByaoZEulso8CqVcg0zNZCey6n2PNMAtUoHAqboG1EmwAciTa49M8qNAAJaTK1QkRBKMkAKg2GDNYGwirgTkIcGIMEc+2U3QQvdoRrAlfrfFr7CQMC4qGb2ZI6Ht2h8AGkwN1qHwiBhnSh1MDtEGdODyhbrbbiPazOVTMtT-efx8GAp6rA466bkg267igx7-veqzIvspglEhagrgAajASAkEQAHqDShT9JU-CsHAPD4CRaDqORmBIHOcY1mgEFNoqLbqAAkm4NIvJgXAAFYODAfDqD4d7qEJImfCi5jlIQWL4MBfH8cs4nrCGuKDNeKAPHwxDghCymqQhYAaUYfBbvsJT7AuS7YKwfBkJQJAkkxLFIPuKrqMeZ4XjEKDXre96PsiLzlMpDgnuU4kOKeXjKHqykGqukp0BBW47nubEHkeJ4gWhtkqiODlOS5wEMMxDasVmXn2uel4BTeYBYTheH3qF0X3rF8V6seflXk1FjGQRiQgasYzCQRqFhaYEU6D1MlJf8KX0OlUGZSgnnOblZ4FVAi5FfZjnOa5FXuVt6gvLc12EFYknqGs0k-qY8msIpYBmrcCBvfgl1nVVHnZV5XHID4t2UJ8VlCqSaLiqlMrDLwlwMvcjzPC8-0btV7EvAARPmOS458kPEjDCbJtclYIxEuaRumiaFtywC8vyuRAA) with a couple contrived examples of the type of predicates I'm hoping to add helpers for.~~ edit: see next comment for playground

**Comment by airjp73** [comment]
I was playing around with this idea a bit more and I think three helpers would be enough to cover any case I could think of.

* `hasProp` - simply determines if an object has the given prop and narrows the type accordingly
* `hasPropSatisfying` - everything `hasProp` does but also checks that the prop satisfies the provided predicate
* `shallowEquals` - Does a shallow `===` comparison. Useful for narrowing discriminated unions.

I implemented a proof of concept for this in [this typescript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEpwGZQiOAcigFMQSATZQgbgCg6YBPMEuAFQgojgF446cOAG84qGMhgBXVAC4iAY3xgANiRiVCAGjhLwajRQCCMeQBFJbAD5iYUYADsA5jo0APU7fvOdUsFUMAISYAfnlxbyc4AF9BOBtRcUkZeUIHCBgAWj1VdU1XEg9wu0cXOD8AymCwr1KYuiUHcTgYLgh0fgBtAF0UdE5uHoZmVjgjFRUAVQdgCAcAaRImVAAeadmHAD4+OHW5uEKNBwp0PYc4kLgAayWIbF2Zubj5BxIANxIoOMYWNgBZZA3RZMDAkACOUmApAoKwA8gAjABWOgACngwAcPCRjugbkw7nAEYjtrw4kS4AAyTAkJRQGFoiBgHQAUTcChUUgoJDhSM6DLA3V8xxI2EclE2mx+owA6sAYAALUEQqGUfk85FwfmYo4nMYTM7A1ZEiV8OIAoFLJWQ6ErVl2ZAKGDqnSiTp4uCOTXo7o1ZAOJgxTao9GbegjNiyhX82HYdi-Z1exna7G6vEEonBxlx0aHFPoIl870kkRxN1LD0Oa63e5E7ryd25nGJjGXO1QB1Ogt4wXN7MkbbyLtLbp0WIMAD047gAAkSCpWFBUHRJ3BpRApCoKCgKFuAshMqLF-APouNug0HAAO5zlQMbBSByOjZweVofkAZUkwFQ2CYpQTWqNrq4xTI8CxLEaSJBmM2AaFAaKUMACiWMmTaRoq4LWqq6IASGhaMt0mwABRxHi8j8locRgNCSGWPIRHUYyg68vy3QAJR8NsjEYt+MFwQhFC0RodAccIcSkNIUCVkREBIsxiJsfIsmIh66DodGsbxuhVoqhQaoZs20EUXxnwCUJ-acXA1GIchGgyby3ZsaOd4Pk++yvqg+lIqhurKTSMCZhiQHoCBBoQeqEokUIZHNiJJZCBJUhSXA9mIvJilwMpqmrnKGHKtCXkavyJLbO6nrKc5dD3o+MDPqgr4TBAl7MhCyAqKsspcn2OiNM0ABqbVSGwwU5V1vzEW8g0kPIA0ciQonieoSXSZewBcq8FDyJ1JB9hlq3rZQ2WzUNln7dih28JdcCTXN9Bjiu7AkOIS6MG0S5CAAdKKKhwURHkfl+P5-s4RGEBUlgUME2iYB934AHIZERGCw6gcPACobGY6WAAM3QfeDQRMPQQgrgAeiEwxvV96O-f96KgwTVRMIQWNCJ0uP4-4EPBMuU5CHA5MML18CrQqADCygGJQJg7K03CoNTP2fH9b704QORS8YMAs2x9DC7YyQdFeuUS-oeRa+zeNJNIqD0Cu-OC-rGvmzL-Ci-Kpu5IYJiWx9zvezAvP8wLFMNHMzTu57msmBYGiy1T3206rjKfrVQOlKD-vS9rOjI9+sfzU5+vWzIABMOyR5LLswAXvsl7bQcOxTTtVwH5duybrfZ7XHNZ1rjdCILYdNCLndm4Y74SDb8fy4rSeeeiqffr+GcAET16vOj1W1KhNS1UhtagRGr33q+Y05bMc-Xdt84PIRAA).

**Comment by eranhirsch** [comment]
Hey, sorry for the long delay in getting back to you. I'm supportive of all of these if you want to send a PR to add them.

`shallowEquals` should be named `isEquals`. We need to deprecate the existing `equals` function and rename it `isDeepEquals` in a later release of v2.

**Comment by TkDodo** [comment]
> shallowEquals should be named isEquals

hmm whats the reasoning here? Isn't shallowEquals and deepEquals an established naming pattern?

anyway, I agree with deprecating equals because it's unclear what it does and we could just make it an alias of the deep equals function

**Comment by TkDodo** [comment]
isEqual and isDeepEqual are also fine if we want functions that return a boolean to start with is or has. But I've never seen `isEquals`...

**Comment by eranhirsch** [comment]
I'm very pedantic, so I like the convention of isXXX for boolean functions; I like it so much that I even prefer to avoid hasXXXX when I can and consider the isXXX convention a syntactic requirement. At facebook all async functions had to be called `genXXXX`, as in `async function genFoo()` and there were lint rules (maybe even compiler warnings) to protect this.

**Comment by eranhirsch** [comment]
@TkDodo  - just realized you were referring to the plural 's' at the end of `isEquals` and not the `is` part. I **totally** agree on that! 

To summarize, we should have an `isEqual` (for shallow equality checks) and `isDeepEqual`.

**Comment by airjp73** [comment]
No worries on response times. I totally get it! 😁

I'll see if I can get a PR up this week.

**Comment by TkDodo** [comment]
> To summarize, we should have an isEqual (for shallow equality checks) and isDeepEqual.

could we explicit and do `isShallowEqual` and `isDeepEqual` ? 

But looking at the PR, it says that:

> isEqual does a shallow, Object.is equality comparison and acts as a type guard

`Object.is` is an equal comparison, but it's not "shallow". What is commonly understood as shallow equality is that objects and arrays are compared one level deep, so `isShallowEqual({ a: 'foo' }, { a: 'foo' }` would be true.

`Object.is` does a referential equality comparison. So do we then need 3 functions?

**Comment by airjp73** [comment]
Good point of clarification. I misunderstood "shallow equality" to be essentially a synonym for "referential equality". I'm primarily interested in comparing primitives so I'm down for whichever approach you both think is best.

**Comment by TkDodo** [comment]
my suggestion would be:

- `isEqual` - does a referential comparison with `Object.is`
- `isShallowEqual` - does a shallow equal by comparing one level deep, e.g. [how zustand does it](https://github.com/pmndrs/zustand/blob/6109bc3bd0f3850c2d9546956de971f27834ac7a/src/vanilla/shallow.ts)
- `isDeepEqual` - deep equals comparison, like we do now with `equals`

also:
- deprecate `equals` and basically alias it for `isDeepEqual`

@eranhirsch is that alright ?

**Comment by eranhirsch** [comment]
Sounds good!

**Comment by eranhirsch** [comment]
Well... one point, about discoverability, If we put everything under a single namespace it's easier for people to find it, `isEqual.shallow`, `isEqual.deep`, and `isEqual`, otherwise they don't share a prefix and typeahead wouldn't show the other... I don't think it's necessarily a better API, but we can talk about it and see how we feel.

**Comment by TkDodo** [comment]
disadvantage is that those wouldn't be tree-shakable, and we don't have those "namespaces" anywhere else. The only thing this is currently used for is for `.strict`, and `.strict` doesn't have a separate runtime implementation - only difference is type-level, so the tree-shaking problem doesn't apply. I don't think we should open this box to have different impls on "namespaces" ?

**Comment by eranhirsch** [comment]
The other alternative is `isEqualShallow` `isEqualDeep` and `isEqual`...

**Comment by TkDodo** [comment]
if the issue is discoverability, I think searching for `equal` should yield all 3 results. algolia docsearch would  help here
